utxonursery.go

```diff
@@ -672,123 +653,44 @@ func (u *utxoNursery) reloadPreschool() error {
 
 // reloadClasses reinitializes any height-dependent state transitions for which
 // the utxonursery has not received confirmation, and replays the graduation of
-// all kindergarten and crib outputs for heights that have not been finalized.
+// all kindergarten and crib outputs for all heights up to the current block.
 // This allows the nursery to reinitialize all state to continue sweeping
-// outputs, even in the event that we missed blocks while offline.
-// reloadClasses is called during the startup of the UTXO Nursery.
-func (u *utxoNursery) reloadClasses(lastGradHeight uint32) error {
-	// Begin by loading all of the still-active heights up to and including
-	// the last height we successfully graduated.
-	activeHeights, err := u.cfg.Store.HeightsBelowOrEqual(lastGradHeight)
+// outputs, even in the event that we missed blocks while offline. reloadClasses
+// is called during the startup of the UTXO Nursery.
+func (u *utxoNursery) reloadClasses(bestHeight uint32) error {
+	// Loading all active heights up to and including the current block.
+	activeHeights, err := u.cfg.Store.HeightsBelowOrEqual(
+		uint32(bestHeight))
 	if err != nil {
 		return err
 	}
 
-	if len(activeHeights) > 0 {
-		utxnLog.Infof("Re-registering confirmations for %d already "+
-			"graduated heights below height=%d", len(activeHeights),
-			lastGradHeight)
+	// Return early if nothing to sweep.
+	if len(activeHeights) == 0 {
+		return nil
 	}
 
+	utxnLog.Infof("(Re)-sweeping %d heights below height=%d",
+		len(activeHeights), bestHeight)
+
 	// Attempt to re-register notifications for any outputs still at these
 	// heights.
 	for _, classHeight := range activeHeights {
-		utxnLog.Debugf("Attempting to regraduate outputs at height=%v",
+		utxnLog.Debugf("Attempting to sweep outputs at height=%v",
 			classHeight)
 
-		if err = u.regraduateClass(classHeight); err != nil {
-			utxnLog.Errorf("Failed to regraduate outputs at "+
+		if err = u.graduateClass(classHeight); err != nil {
+			utxnLog.Errorf("Failed to sweep outputs at "+
 				"height=%v: %v", classHeight, err)
 			return err
 		}
 	}
 
-	// Get the most recently mined block.
-	_, bestHeight, err := u.cfg.ChainIO.GetBestBlock()
-	if err != nil {
-		return err
-	}
-
-	// If we haven't yet seen any registered force closes, or we're already
-	// caught up with the current best chain, then we can exit early.
-	if lastGradHeight == 0 || uint32(bestHeight) == lastGradHeight {
-		return nil
-	}
-
-	utxnLog.Infof("Processing outputs from missed blocks. Starting with "+
-		"blockHeight=%v, to current blockHeight=%v", lastGradHeight,
-		bestHeight)
-
-	// Loop through and check for graduating outputs at each of the missed
-	// block heights.
-	for curHeight := lastGradHeight + 1; curHeight <= uint32(bestHeight); curHeight++ {
```

Comment:
Nice, this should speed up the start up after we've been offline for a long-ish time as well.